XMonad/Layout/Columns.hs

```diff
@@ -0,0 +1,478 @@
+-- > ((modm .|. controlMask, xK_j), sendMessage VerticalShrink),
+--
+-- This layout is known to work with:
+--
+-- * "XMonad.Layout.WindowNavigation" for changing focus with a direction using
+-- 'XMonad.Layout.WindowNavigation.Go' messages.
+-- * 'XMonad.Layout.SubLayouts.subTabbed' for docking windows together with
+-- tabs. Note that sometimes when undocking windows, the layout is reset. This is
+-- a minor annoyance caused by the difficulty to track windows in the sublayout.
+
+-- | The windows can be moved in every directions.
+--
+-- Horizontally, a window alone in its column cannot be moved before the first
+-- or after the last column. If not alone, moving the window outside those
+-- limits will create a new column.
+-- The windows can also be moved vertically in their column.
+data Move = MoveLeft | MoveRight | MoveUp | MoveDown deriving (Show, Read)
+
+instance Message Move
+
+-- | The windows can be resized in every directions.
+--
+-- When resizing horizontally:
+--
+-- * if the window to be resized is not in the last column
+--
+--      * then the right side of the window will be moved
+--      * the last column will compensate the size change
+--
+-- * if the window is in the last column
+--
+--      * then the left side of the window will be moved
+--      * the column on the left of the current one will compensate the size change
+--
+-- The same applies when resizing vertically using the bottom side of the
+-- window unless it is the last window in the column in which case we use the
+-- top side.
+data Resize
+  = VerticalShrink
+  | VerticalExpand
+  | HorizontalShrink
+  | HorizontalExpand
+  deriving (Show, Read)
+
+instance Message Resize
+
+-- | The layout handles focus change messages.
+--
+-- Built-in focus cannot be used here because @XMonad@ does not make it easy to
+-- change the order of windows in the focus list. See also 'focusUp' and
+-- 'focusDown' functions.
+data Focus = FocusUp | FocusDown
+  deriving (Show, Read)
+
+instance Message Focus
+
+-- | A column is a list of windows with their relative vertical dimensions.
+type Column = [(Rational, Window)]
+
+-- | The layout is a list of 'Column' with their relative horizontal dimensions.
+type Columns = [(Rational, Column)]
+
+newtype ColumnsLayout a = Columns Columns
+  deriving (Show, Read)
+
+instance LayoutClass ColumnsLayout Window where
+  description _ = layoutDescription
+
+  emptyLayout _ _ = pure ([], Just $ Columns [])
+
+  doLayout (Columns columns) rectangle stack =
+    pure (rectangles, Just (Columns columns'))
+    where
+      hackedColumns = hackForTabs columns stack
+      columns' = updateWindowList hackedColumns stack
+      rectangles = toRectangles rectangle' columns'
+      -- If there is only one window and the screen is big, we reduce the
+      -- destination rectangle to put the window on the center of the screen.
+      rectangle'
+        | rect_width rectangle > 2000 && (length . toList $ stack) == 1 =
```

Comment:
I would agree that this should be configurable. You may also want to look into [XMonad.Layout.PerScreen](https://hackage.haskell.org/package/xmonad-contrib-0.18.0/docs/XMonad-Layout-PerScreen.html) (but see #813); perhaps there's a way to farm this decision out to users' configs, allowing them to decide if they even want the behavior.

Reply:
Wow. Two weeks with a broken network. Sounds terrible :fearful: 

I wasn't aware of that `PerScreen` module. Thx. Indeed, with it, that feature becomes irrelevant. 

I've made the width of the first column configurable. 

I can now reproduce the old behavior using `PerScreen` with two instances of `Columns` and a different value for that first column width. That's fine for my use case.